tz/tzinit.8
.TH TZINIT 8 "25 April 1999" "MiNTLib Tools"
.SH NAME
tzinit \- initialize kernel time keeping variables
.SH SYNOPSIS
.B tzinit
[
.B OPTIONS
]
.SH DESCRIPTION
.B Tzinit
sets or inquires the FreeMiNT kernel time-keeping variables.  Run with
no options, 
.B tzinit 
will display the contents of all relevant kernel time keeping
variables.  Given the appropriate options will make
.B tzinit 
change a subset of these variables.
.PP
.B Tzinit 
will examine the environment to analyse the settings for the
local timezone.  This will be done with the facilities provided by the
standard programming library.  The MiNTLib and the GNU libc will do
this by analysing the contents of the environment variable
.I TZ
(see
.B environ(5)
for details).  If this variable is not set (which is recommended)
a system-wide default timezone will be read from the file
.I /usr/etc/zoneinfo/localtime
to find out the offset to Greenwich Mean Time for the local timezone.
If set to an empty value, the time zone "GMT" (Greenwich Mean Time)
is assumed.  If the contents of 
.I TZ 
starts with a colon the rest of the contents of the variable is interpreted
as the name of a file to be examined for time zone settings instead of
.I /usr/etc/zoneinfo/localtime.
If none of the above applies the variable is parsed according to the 
syntax described by POSIX.  If no rules for summertime (also known as
Daylight Savings Time DST) were supplied a system-wide default file
.I /usr/etc/zoneinfo/posixrules
is examined. 
.PP
The program should be invoked in the startup-sequence of the FreeMiNT
system, as early as possible because it may cause the kernel clock to
make leaps which can lead to undesirable effects (this can actually
only occur if you use
.B tzinit
to make your kernel clock tick in local time).  It is probably best
to run
.B tzinit
from your
.B mint.cnf(5)
file.
.SH OPTIONS
.PP
These options are available:
.TP
.B \-h, \-\-help
Print a short usage summary and exit.
.TP
.B \-l, \-\-localtime
At system startup the FreeMiNT kernel runs in
.I UTC
.I clockmode.
That means that the time value supplied by the hardware clock facilities is
interpreted as a time in Universal Time Coordinated (UTC).  Given 
.B tzinit
the option
.I \-\-localtime
will change the kernel's behavior so that hardware date and time is interpreted
as a time value in local representation instead.  This facility is useful
when running other old-fashioned operating systems beside FreeMiNT on the same
machine which always interprete the hardware time as a local time.
.TP
.B \-s, \-\-silent, \-\-quiet
This option disables all diagnostic output except error messages.
.TP
.B \-t TIMEZONE, \-\-timezone=TIMEZONE, \-\-tz=TIMEZONE
By default the
.B tzinit
program will inquire the current settings for the local timezone by 
examining the environment variable TZ.  This option lets you override
the value given in the environment.  
.TP
.B \-n, \-\-no-timezone, \-\-no-tz
Removes an existing definition for the variable TZ from the environment.
This option may be useful in situations where you can't unset 
environment variables (like when running from
.I mint.cnf
or the like).  Note that unsetting the TZ environment variable has not
the same effect as setting it to the empty string (see above for details).
.TP
.I 
.TP
.B \-u, \-\-utc, \-\-gmt
Set (or reset) the kernel clock mode to the default (hardware time is
interpreted as Universal Time Coordinated UTC).
.TP
.B \-V, \-\-version
Display a short version information and exit.
.TP
.B \-v, \-\-verbose
Enable verbose diagnostic output (default).  The
.B tzinit
program will display the contents of all relevant kernel time-keeping
variables.
.SH FILES
.PP
.I /usr/share/zoneinfo/localtime
.PP
.I /usr/share/zoneinfo/posixrules
.SH AUTHOR
The 
.B tzinit
program was written by Guido Flohr <gufl0000@stud.uni-sb.de>.  It is part
of the MiNTLib project and available under the COPYMINT license.
.SH STANDARDS
Pure anarchy, no standard knows this program.
.SH "SEE ALSO"
Tsettimeofday(2), Ssystem(2), tzset(3), tzsetwall(3), ctime(3), 
environ(5), tzfile(5), zic(8), zdump(8)


NEWS
Version 0.55.3 - 07 May 2000

* Fixed st_blocks in stat() and friends.

* New and quicker qsort algorithm.

* Changed time_t to long int instead of int.

Version 0.55.2 - 10 April 2000

* Added forgotten include/bits/mint_stdio.h.

Version 0.55.1 - 09 April 2000

* Minor bugfixes.

* libc-symbols.h not installed any more.

* Missing functions added: llabs, strtoimax, strtoumax.

* argz functionality.

Version 0.55 - 02 April 2000

* Countless bugfixes.

* This version of the MiNTLib requires gcc 2.95.2, older gcc versions
  will not work, you can neither compile nor use the MiNTLib without
  them.  See README.1st for details.
  
* Introduction of a new stdio model, mostly taken from the GNU libc.
  The new stdio API is incompatible with previous releases.  See
  README.stdio to check if you have to recompile your binaries
  The new stdio model provides a lot of new features, see bits/mint_stdio.h
  for details.

* The data type long double has changed.  It now complies with 
  IEEE854, which will require recompilation of libraries or object 
  files that do long double arithmetics.  Accuracy problems with other 
  floating  point types have been solved.
  
* The source tree has been completely restructured.

* The MiNTLib now comes along with an exhaustive test suite.  Run
  "make -k check" in the compiled source tree to see the results.
  See README.tests for details.

* UFC (ultra-fast crypt) and support for MD5-crypt.  The new crypt()
  resp. fcrypt() routines outperform previous MiNTLib releases
  by factor 5 here.

* Incompatible change in select(): The function now internally calls
  poll().  The data type fd_set has changed and now has a theoratical
  limit of 1024 file descriptors.  Until the kernel gets changed you're
  still stuck with 32 descriptors though.  However, object files or
  libraries that reference select() have to get recompiled.

* This is the last MiNTLib that builds 16 bit integer libraries by
  default.  See README.mshort for details.

  In previous (unofficial) releases:
  
* The former portlib and the socketlib from the MiNTNet package have 
  been integrated into the MiNTLib.  

* The format of struct stat in <sys/stat.h> has changed in an 
  incompatible way -> recompile.

* Shadow password and group database.

* New management functions for utmp/wtmp.

* The library is now installed in /usr instead of /usr/local by
  default.



tz/tzinit.c
/*  tzinit.c -- MiNTLib.
    Copyright (C) 1999 Guido Flohr <guido@freemint.de>

    This file is part of the MiNTLib project, and may only be used
    modified and distributed under the terms of the MiNTLib project
    license, COPYMINT.  By continuing to use, modify, or distribute
    this file you indicate that you have read the license and
    understand and accept it fully.
*/

#ifndef _GNU_SOURCE
# define _GNU_SOURCE
#endif

#include <stdio.h>
#include <stdlib.h>
#include <errno.h>
#include <sys/time.h>
#include <time.h>
#include <string.h>

#include <getopt.h>
#include <error.h>

#include <mint/mintbind.h>
#include <mint/ssystem.h>

static int verbose = 1;

static const struct option long_options[] =
{
  { "gmt", no_argument, NULL, 'u' },
  { "help", no_argument, NULL, 'h' },
  { "localtime", no_argument, NULL, 'l' },
  { "no-timezone", no_argument, NULL, 'n' },
  { "no-tz", no_argument, NULL, 'n' },
  { "quiet", no_argument, NULL, 's' },
  { "silent", no_argument, NULL, 's' },
  { "tz", required_argument, NULL, 't' },
  { "timezone", required_argument, NULL, 't' },
  { "utc", no_argument, NULL, 'u' },
  { "verbose", no_argument, NULL, 'v' },
  { "version", no_argument, NULL, 'V' },
  { 0, 0, 0, 0 }
};

static void show_help __PROTO ((void));
static void show_version __PROTO ((void));
static void show_usage __PROTO ((void));

static int do_utc = 0;
static int do_local = 0;
static int no_tz = 0;
static int do_tz = 0;
static char* newtz = NULL;

int main (argc, argv)
    int argc;
    char* argv[];
{
  int the_option;
  struct timezone tz;
  struct tm* tm;
  time_t now;
  
  while ((the_option = getopt_long (argc, argv, "hlst:uVv", 
                                    long_options, NULL)) != EOF)
    {
      switch (the_option)
        {
          case 0:  /* Long option that just sets a flag.  */
            break;
      
          case 'h':
            show_help ();
            break;
        
          case 'l':
            do_local = 1;
            if (do_utc)
              {
                error (EXIT_SUCCESS, 0, ("\
The options `-u' and `-l' are mutually exclusive."));
                show_usage ();
              }
            break;
              
          case 's':
            verbose = 0;
            break;
          
          case 't':
            do_tz = 1;
            if (no_tz)
              {
                error (EXIT_SUCCESS, 0, ("\
The options `-n' and `-t' are mutually exclusive."));
                show_usage ();
              }
            newtz = optarg;
            break;
            
          case 'n':
            no_tz = 1;
            if (do_tz)
              {
                error (EXIT_SUCCESS, 0, ("\
The options `-n' and `-t' are mutually exclusive."));
                show_usage ();
              }
            if (putenv ("TZ") != 0)
              error (EXIT_FAILURE, errno, "can\'t unset TZ envariable");
            break;
            
          case 'u':
            do_utc = 1;
            if (do_local)
              {
                error (EXIT_SUCCESS, 0, ("\
The options `-u' and `-l' are mutually exclusive."));
                show_usage ();
              }
            break;
              
          case 'V':
            show_version ();
      
          case 'v':
            verbose = 1;
            break;
            
          default:
            show_usage ();
        }
    }
  
  /* Check if Ssystem is available.  */
  __set_errno (-Ssystem (-1, 0, 0));
  if (errno != 0)
    error (EXIT_FAILURE, errno, ("Ssystem kernel facility not available"));
  
  /* If timezone was set on the command line, set it now.  */
  if (newtz != NULL)
    {
      char* tzenv = malloc (strlen (newtz) + 4);
      
      if (tzenv == NULL)
        error (EXIT_FAILURE, 0, "virtual memory exhausted");

      sprintf (tzenv, "TZ=%s", newtz);
      
      if (putenv (tzenv) != 0)
        error (EXIT_FAILURE, errno, "can\'t set TZ environment variable");
        
      tzset ();
    }
  
  if (do_utc || do_local)
    {
      int clockmode = do_utc ? 0 : 1;
      
      /* Now set the kernel timezone.  */
      now = time (NULL);
      if (now == (time_t) -1)
        {
          /* It is not necessarily sufficient to check if time returns -1
             because that is also a valid return value for Dec 31, 1969,
             23:59:59.  But we are in the nineties now and if we get this
             value there has something necessarily gone wrong.  */
           error (EXIT_FAILURE, errno, "can\'t get current time.\n");
        }
  
      tm = localtime (&now);
      if (tm == NULL)
          error (EXIT_FAILURE, errno, "can\'t convert to localtime.\n");

      tz.tz_minuteswest = tm->TM_GMTOFF / (-60);
      tz.tz_dsttime = tm->tm_isdst;
  
      if (Tsettimeofday (NULL, &tz) != 0)
        error (EXIT_FAILURE, errno, "settimeofday failed");

      /* Department of unclean tricks (setting errno).  */
      __set_errno (-Ssystem (S_CLOCKUTC, clockmode, 0));
      if (errno != 0)
        error (EXIT_FAILURE, errno, "can\'t set kernel clock mode");
      __set_errno (0);
  }
  
  /* If nothing to do, display current settings.  */
  if (verbose)
    {
      time_t now = time (NULL);
      struct tm* tm = localtime (&now);
      int gmtoff = tm->TM_GMTOFF;
      int sec_east = gmtoff % 60;
      int hrs_east = tm->TM_GMTOFF / 3600;
      int min_east = 60 * (gmtoff - sec_east - 3600 * hrs_east);
      
      printf ("Current date and time: %s", asctime (tm));
      printf ("Time zone in use: %s\n", tm->TM_ZONE);
      printf ("East of Greenwich Mean Time: %d:%02d:%02d\n", 
          hrs_east, min_east, sec_east);
      printf ("Kernel clock mode: %s\n", Ssystem (S_CLOCKUTC, -1, 0) ? 
              "localtime" : "UTC");
    }
    
  return EXIT_SUCCESS;
}

static 
void show_help ()
{
  printf ("\
Usage: %s [OPTIONS] ...\n\
Arguments that are mandatory to short options are mandatory to long\n\
options too.\n\
\n\
  -h, --help             display this help page and exit\n\
  -l, --localtime        set kernel clock to localtime mode\n\
  -s, --silent, --quiet  don\'t display any informative messages\n\
  -t, --timezone=TIMEZONE, --tz=TIMEZONE\n\
                         set timezone to TIMEZONE (instead of reading\n\
                         the environment)\n\
  -n, --no-timezone, --no-tz
                         remove an eventually existing setting for the\n\
                         variable TZ from the environment
  -u, --utc, --gmt       set kernel clock mode to UTC mode\n\
  -V, --version          display version information and exit\n\
  -v, --verbose          display verbose information (default)\n\
\n\
Set or examine kernel time-keeping variables.\n", program_invocation_name);

  exit (EXIT_SUCCESS);
}

static
void show_version ()
{
  if (program_invocation_short_name == NULL 
      || program_invocation_short_name[0] == '\0')
    program_invocation_short_name = "tzinit";
    
  if (__MINTLIB_REVISION__ != 0)
    printf ("%s - MiNTLib version %d.%d.%d\n", program_invocation_short_name,
        __MINTLIB_MAJOR__, __MINTLIB_MINOR__, __MINTLIB_REVISION__);
  else
    printf ("%s - MiNTLib version %d.%d.%d\n", program_invocation_short_name,
        __MINTLIB_MAJOR__, __MINTLIB_MINOR__, __MINTLIB_REVISION__);
  
  printf ("\
Copyright (C) 1999 Guido Flohr <guido@freemint.de>\n\
This program is free software, see the sources for copying conditions.\n\
There is no warranty, not even for merchantibility or fitness for a\n\
particular purpose.\n");
  exit (EXIT_SUCCESS);
}

static
void show_usage ()
{
  fflush (stdout);
  fprintf (stderr, "Try `%s --help' for more information.\n",
      program_invocation_name);
  exit (EXIT_FAILURE);
}
